flush/sink.go

```diff
@@ -0,0 +1,152 @@
+
+// WithAckFunc set the AckFunc callback function, which is called for each ack recieved. If the
+// AckFunc returns an error, it will be returned by `Run`.
+func WithAckFunc(ackFn AckFunc) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackFn = ackFn
+	}
+}
+
+// WithMsgBufferSize sets the msg channel buffer size.
+func WithMsgBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.msgBufferSize = size
+	}
+}
+
+// WithAckBufferSize sets the ack channel buffer size.
+func WithAckBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackBufferSize = size
+	}
+}
+
+// AckFunc is a callback function executed by `Run` for each message succesfully produced. The ack
+// counter is only incremented once this function returns. If no AckFunc is provided the ack counter
+// is always incremented.
+type AckFunc func(msg substrate.Message) error
+
+// AsyncMessageSink wraps substrate.AsyncMessageSink and provides an interface for interaction with
+// the underlying sink, as well as the capability to flush messages.
+type AsyncMessageSink struct {
+	sink          substrate.AsyncMessageSink
+	msgs          uint64
+	msgBufferSize int
+	msgCh         chan substrate.Message
+	acks          uint64
+	ackBufferSize int
+	ackCh         chan substrate.Message
+	ackFn         AckFunc
+}
+
+// NewAsyncMessageSink returns a pointer a new AsyncMessageSink.
+// See examples/flush/sink.go for example usage.
+func NewAsyncMessageSink(sink substrate.AsyncMessageSink, opts ...AsyncMessageSinkOption) *AsyncMessageSink {
+	ams := &AsyncMessageSink{
+		sink:          sink,
+		msgBufferSize: defaultMsgBufferSize,
+		ackBufferSize: defaultAckBufferSize,
+	}
+
+	for _, opt := range opts {
+		opt(ams)
+	}
+
+	ams.msgCh = make(chan substrate.Message, ams.msgBufferSize)
+	ams.ackCh = make(chan substrate.Message, ams.ackBufferSize)
+
+	return ams
+}
+
+// Run initialises message publishing using the underlying sink and blocks until either an error
+// occurs or the context is done. If an AckFunc is configured, Run will execute it for each ack
+// recieved. If an error is returned, the user should cancel the context to prevent `Flush` from
+// blocking.
+func (ams *AsyncMessageSink) Run(ctx context.Context) error {
+	group, groupctx := rungroup.New(ctx)
+
+	group.Go(func() error {
+		return ams.sink.PublishMessages(groupctx, ams.ackCh, ams.msgCh)
+	})
+
+	group.Go(func() error {
+		for {
+			select {
+			case <-groupctx.Done():
+				return nil
+			case msg := <-ams.ackCh:
+				if ams.ackFn != nil {
+					err := ams.ackFn(msg)
+					if err != nil {
+						return err
+					}
+				}
+
+				atomic.AddUint64(&ams.acks, 1)
+			}
+		}
+	})
+
+	return group.Wait()
+}
+
+// PublishMessage publishes a message to the underlying sink. PublishMessage is desgined to be
+// called concurrently by the user.
+func (ams *AsyncMessageSink) PublishMessage(msg []byte) {
```

Comment:
can this be the substrate message interface rather than a simple byte slice?

Reply:
I did initially use `substrate.Message` here, instead of `[]byte`. However I thought it better to use `[]byte` and convert to a `substrate.Message` type within the `PublishMessage` function using `message.NewMessage`, rather than having the user do it. That way we avoid this issue: https://github.com/uw-labs/substrate/issues/103.

With that said, I'm easy either way. Accepting a `substrate.Message` would be more consistent with substrate itself and other tooling. However, 99% of the time the the user would simply convert a `[]byte` to `substrate.Message` anyway - and this way guarantees serialisation (or, at least makes it not the responsibility of the user).